### PR TITLE
Add graphic manager, tweak Tileset Editor and Room Config Dialog to support editing layer 3

### LINKED ISSUE #211: blank Layer render problem
background layers will have weird things from other Rooms' layers or even non-map layers sometimes due to Qt internal memory problems perhaps.
image

### LINKED ISSUE #225: Open a second rom will make the editor unstable and crash or cause corruption in the rom file

### LINKED ISSUE #273: Support importing & deleting background tileset in Tileset editor

### LINKED ISSUE #274: Support editing and saving mapping type 0x2X layers using background tiles
some extra ideas:
we should support editing layers using not-in-the-rom tileset too, so users can make some cutscenes by themselves using patch.
support exporting raw, RLE compressed, Lz77 compressed layer data so we can use them in patches.

### LINKED ISSUE #275: Support selecting mapping type 0x2X layers in Room Config Dialog
support selecting new created mapping type 0x2X layers, but not only the vanilla one which has been hardcode in the editor.
the only problem here is that, we cannot list all the layers which can be used by the modified background tile set.
keep the old hardcode layers and add in the new one existing in the Rooms should be easy, just loop through all the Levels and Rooms, but there can be some unused layers exist in the rom data. also if 2 background tile sets are similar, can we cross-referance their existing layers? and another problem is what if some background tile set get modified, can those existing layers still works fine or we need to delete some because they are useless after the modification on the tileset? if so, we need to check if they are cross-referred too to see if they can be deleted, and this should be decided by users.

### LINKED ISSUE #376: some problems found in OAM Designer
users are able to edit the string in the list widget by double click the item
character id should be in hex, or at least compatible with the tile id when selecting tile in the sprite graphic panel
attempt to create oam for some entity like treasurebox (when global sprite id is 1) will cause the whole software crash
cannot delete the last oam in the data list widget
do we need to add options to offset the palette or tile id ?

### PR DESCRIPTION
**Pull request by shinespeciall**
Fix #211 
Resolve #274 
Resolve #273 
Resolve #275 
Fix #376 
Resolve #225

### CODE REVIEW

File: LevelComponents/Room.h [hunk 1]
```diff
@@ -212,6 +214,7 @@ namespace LevelComponents
         void DeleteEntity(int index);
         void DeleteEntity(int difficulty, int index);
         void ClearEntitylist(int difficulty);
+        void ResetRoomHeaderAddr(unsigned int newaddr) { headerAddr = newaddr; }
```
Comment (Goldensunboy) [inline LevelComponents/Room.h:217]: should be called Setxxxxxxxxxxx not Reset

File: Operation.cpp [hunk 2]
```diff
@@ -244,6 +247,9 @@ void BackTrackOperation(struct OperationParams *operation)
         singleton->SetEditModeDockWidgetLayerEditability();
         singleton->SetEditModeWidgetDifficultyRadioBox(1);
         singleton->ResetEntitySetDockWidget();
+
+        // hint to show undo operation
+        singleton->GetOutputWidgetPtr()->PrintString(QObject::tr("Undo Roomo Config changes."));
```
Comment (Goldensunboy) [inline Operation.cpp:252]: roomo

File: Dialog/GraphicManagerDialog.cpp [hunk 3]
```diff
@@ -0,0 +1,1741 @@
...
+    ui->comboBox_mappingDataType->addItems(GraphicMappingDataCompressionTypeName);
+    ui->graphicsView_tile8x8setData->scale(2, 2);
+    ui->graphicsView_palettes->scale(2, 2);
+    ui->graphicsView_mappingGraphic->scale(1, 1);
+
+    if (graphicEntries.size())
+    {
+        graphicEntries.clear();
+    }
+    graphicEntries = ScatteredGraphicUtils::GetScatteredGraphicsFromROM();
```
Comment (Goldensunboy) [inline Dialog/GraphicManagerDialog.cpp:43]: Can we change all instances of "scattered" to "assorted"

Comment (shinespeciall) [reply Dialog/GraphicManagerDialog.cpp:43]: done

**Review by chanchancl — APPROVED** [review]

**Pull request merged**